test/e2e/helpers/toxiProxy.go

```diff
@@ -0,0 +1,188 @@
+package helpers
+
+import (
+	"context"
+	"fmt"
+	. "github.com/onsi/gomega"
+	"math/rand"
+	"os"
+	"path"
+	"regexp"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/exec"
+	"strconv"
+	"strings"
+)
+
+func ToxiProxyServerExec(ctx context.Context) error {
+	execArgs := []string{"run", "-d", "--name=capc-e2e-toxiproxy", "--net=host", "--rm", "ghcr.io/shopify/toxiproxy"}
+	runCmd := exec.NewCommand(
+		exec.WithCommand("docker"),
+		exec.WithArgs(execArgs...),
+	)
+	_, stderr, err := runCmd.Run(ctx)
+	if err != nil {
+		fmt.Println(string(stderr))
+	}
+	return err
+}
+
+func ToxiProxyServerKill(ctx context.Context) error {
+	execArgs := []string{"stop", "capc-e2e-toxiproxy"}
+	runCmd := exec.NewCommand(
+		exec.WithCommand("docker"),
+		exec.WithArgs(execArgs...),
+	)
+	_, _, err := runCmd.Run(ctx)
+	return err
+}
+
+func ToxiProxyCli(ctx context.Context, command string, args ...string) (string, error) {
+	// Toxiproxy offers a nice API, but it hasn't published new go packages for it since 2019.
+	// The most recently published package can't be installed because of dependent package
+	// case incompatibility (logrus).  Sigh.
+	// So, we're gonna just go with the cli in the docker image.
+
+	execArgs := append([]string{"run", "--rm", "--net=host", "--entrypoint=/toxiproxy-cli", "-t", "ghcr.io/shopify/toxiproxy", command}, args...)
+	runCmd := exec.NewCommand(
+		exec.WithCommand("docker"),
+		exec.WithArgs(execArgs...),
+	)
+	stdout, stderr, err := runCmd.Run(ctx)
+	if err != nil {
+		fmt.Printf("stdout:\n%v\n\nstderr:\n%v", string(stdout), string(stderr))
+		return "", err
+	}
+	return string(stdout), nil
+}
+
+type ToxiProxy struct {
+	context        context.Context
+	KubeconfigPath string
+	ClusterProxy   framework.ClusterProxy
+	ProxyName      string
+}
+
+func SetupForToxiproxyTesting(ctx context.Context, bootstrapClusterProxy framework.ClusterProxy) *ToxiProxy {
+	// Read/parse the actual kubeconfig for the cluster
+	kubeConfig := NewKubeconfig()
+	unproxiedKubeconfigPath := bootstrapClusterProxy.GetKubeconfigPath()
+	err := kubeConfig.Load(unproxiedKubeconfigPath)
+	Expect(err).To(BeNil())
+
+	// Get the cluster's server url from the kubeconfig
+	server, err := kubeConfig.GetCurrentServer()
+	Expect(err).To(BeNil())
+
+	// Decompose server url into protocol, address and port
+	serverRegex := regexp.MustCompilePOSIX("(https?)://([0-9]+\\.[0-9]+\\.[0-9]+\\.[0-9]+):([0-9]*)")
+	urlComponents := serverRegex.FindStringSubmatch(server)
+	Expect(len(urlComponents)).To(Equal(4))
+	protocol := urlComponents[1]
+	address := urlComponents[2]
+	port, err := strconv.Atoi(urlComponents[3])
+	Expect(err).To(BeNil())
+
+	// Declare a port for the proxy to listen on
+	// ToDo: make this only return an unused port
```

Comment:
I don't love this technique, but haven't yet chosen to invest into probing the proposed ports.

Reply:
You can set the port to zero and let Toxiproxy pick an ephemeral port.  How hard would it be to get the chosen port number from the ToxiProxyCli output?

Reply:
Thanks Mark.  That worked well.  After re-writing to use the API instead of the CLI retrieving the assigned port was easy.